File: src/java/com/leetcode/algorithms/medium/Problem78.java
package com.leetcode.algorithms.medium;

/*
 * 78. Subsets
 * https://leetcode.com/problems/subsets/
 * Tags: Backtracking
 * */

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class Problem78 {

  public List<List<Integer>> subsets(int[] nums) {
    List<List<Integer>> result = new ArrayList<>();
    result.add(new ArrayList<Integer>());
    dfs(nums, result, new ArrayList<Integer>(), 0);
    return result;
  }

  private void dfs(int[] nums, List<List<Integer>> result, List<Integer> currList, int index) {
    if (index == nums.length) {
      return;
    }
    List<Integer> newList = new ArrayList<>(currList);
    newList.add(nums[index]);
    result.add(newList);
    dfs(nums, result, currList, index+1);
    dfs(nums, result, newList, index+1);
  }

  public List<List<Integer>> subsetsIterative(int[] nums) {
    List<List<Integer>> result = new ArrayList<>();
    result.add(new ArrayList<Integer>());
    for(int i=0; i<nums.length; ++i) {
      List<List<Integer>> temp1 = new ArrayList<>();
      for(int j=0;j<result.size(); ++j) {
        List<Integer> temp = result.get(j).stream().collect(Collectors.toList());
        temp.add(nums[i]);
        temp1.add(temp);
      }

      temp1.stream().forEach(l -> {
        result.add(l);
      });
    }
    return result;
  }

  public static void main(String[] args) {
    Problem78 problem78 = new Problem78();
    int nums[] = {1, 2, 3};
    System.out.println(problem78.subsets(nums));//[[],[1],[2],[1,2],[3],[1,3],[2,3],[1,2,3]]
    System.out.println(problem78.subsetsIterative(nums));//[[],[1],[2],[1,2],[3],[1,3],[2,3],[1,2,3]]
  }
}

File: src/js/Problem78.js
/*
    Approach 1:
    DFS approach
*/

let nums = [1,2];

function findSubsets(arr) {
    let subsets = [];
    generateSubsets(0, arr, [], subsets);
    return subsets;
}

function generateSubsets(index, nums, current, subsets) {
    subsets.push([...current]);
    for(let i=index; i<nums.length; ++i) {
        current.push(nums[i]);
        generateSubsets(i+1, nums, current, subsets);
        current.pop();
    }
}

console.log(findSubsets(nums));

/*
    Approach 2: Iterative Approach
    [[]]
    1-> [[], [1]]
    2-> [[], [1], [2], [1,2]]
    3-> [[], [1], [2], [1,2], [3], [1,3], [2,3], [1,2,3]]
*/

function findSubsets(nums) {
    let result = [[]];
    for(let i=0; i<nums.length; ++i) {
        let temp1 = [];
        for(let j=0;j<result.length; ++j) {
            let temp = [...result[j]];
            temp.push(nums[i]);
            temp1.push(temp);
        }
        result.push(...temp1);
    }
    return result;
}

console.log(findSubsets([1,2,3]))